### DISCUSSION TITLE
modification to support golang 1.19

### DISCUSSION THREAD

**Discussion by andypmw** [pashagolub/pgxmock#172] [Q&A]
Thank you for this awesome pgxmock libray...

For some reason, I am not allowed to use the N-2 latest Golang versions on production (1.21 and 1.20).
And my unit-tests that are using pgxmock failed when I run using Golang 1.19.

I have questions:

1. do you have any specific reasons to supports only Golang 1.21 for pgxmock?
2. do you have any idea how to make the pgxmock compatible with Golang 1.19?

Thank you.

**Comment by pashagolub** [comment]
Thanks for your kind words! It means a lot!

> For some reason, I am not allowed to use the N-2 latest Golang versions on production (1.21 and 1.20).

pgxmock is a testing library. It's not supposed to go into production binary. That means you can develop your application with go 1.21 and produce your build with go 1.19 on a build server.

> do you have any specific reasons to supports only Golang 1.21 for pgxmock?

Most of them are connected with the new functionality introduced in [v3 major release](https://github.com/pashagolub/pgxmock/releases/tag/v3.0.0).

> do you have any idea how to make the pgxmock compatible with Golang 1.19?

Well, you are still able to use pre-v3 versions, for example [v2.12](https://github.com/pashagolub/pgxmock/releases/tag/v2.12.0). They all support Golang v1.19.

Just execute this command to force your project to use v2 latest release
```bash
go get github.com/pashagolub/pgxmock/v2
```

**Comment by andypmw** [comment]
Hi @pashagolub ,

thanks for the answer. I am sorry, yes sure the pgxmock will not compiled into production binary.
What I mean was the production CI/CD will execute go test using Go 1.19 before deployment, and my project failed because the pipeline unable to run unit-tests that import pgxmock, that needs Go 1.21.

I have read the pgxmock source code, specifically the `expectations.go` and patched the following original function:
```go
func (e *commonExpectation) fulfilled() bool {
	return e.triggered >= max(e.plannedCalls, 1)
}
```

To this one:
```go
func (e *commonExpectation) fulfilled() bool {
	theMax := uint(1)
	plannedCallsQty := uint(e.plannedCalls)

	if plannedCallsQty > theMax {
		theMax = plannedCallsQty
	}

	return e.triggered >= theMax
}
```

The `max()` global function is introduced on Go 1.21:
https://tip.golang.org/doc/go1.21#:~:text=The%20new%20functions%20min%20and%20max%20compute

By patching the `fulfilled()` function, my unit-tests passed for Go 1.19, Go 1.20, and Go 1.21:
```
andy$ go1.19.13 test ./...
[PASSED]

andy$ go1.20.1 test ./...
[PASSED]

andy$ go1.21.0 test ./...
[PASSED]
```

**Reply by pashagolub** [reply]
Sounds cool. Good to know. Have you considered using pgxmock v2, or do you need new features that came out with v3?
